src/Microsoft.OData.Edm/EdmModelVisitor.cs

```diff
@@ -355,6 +518,56 @@ public void VisitPrimitiveTypeReference(IEdmPrimitiveTypeReference reference)
             }
         }
 
+        public Task VisitPrimitiveTypeReferenceAsync(IEdmPrimitiveTypeReference reference)
+        {
+            switch (reference.PrimitiveKind())
+            {
+                case EdmPrimitiveTypeKind.Binary:
+                    return this.ProcessBinaryTypeReferenceAsync(reference.AsBinary());
+                case EdmPrimitiveTypeKind.Decimal:
+                    return this.ProcessDecimalTypeReferenceAsync(reference.AsDecimal());
+                case EdmPrimitiveTypeKind.String:
+                    return this.ProcessStringTypeReferenceAsync(reference.AsString());
+                case EdmPrimitiveTypeKind.DateTimeOffset:
+                case EdmPrimitiveTypeKind.Duration:
+                case EdmPrimitiveTypeKind.TimeOfDay:
+                    return this.ProcessTemporalTypeReferenceAsync(reference.AsTemporal());
+                case EdmPrimitiveTypeKind.Geography:
+                case EdmPrimitiveTypeKind.GeographyPoint:
+                case EdmPrimitiveTypeKind.GeographyLineString:
+                case EdmPrimitiveTypeKind.GeographyPolygon:
+                case EdmPrimitiveTypeKind.GeographyCollection:
+                case EdmPrimitiveTypeKind.GeographyMultiPolygon:
+                case EdmPrimitiveTypeKind.GeographyMultiLineString:
+                case EdmPrimitiveTypeKind.GeographyMultiPoint:
+                case EdmPrimitiveTypeKind.Geometry:
+                case EdmPrimitiveTypeKind.GeometryPoint:
+                case EdmPrimitiveTypeKind.GeometryLineString:
+                case EdmPrimitiveTypeKind.GeometryPolygon:
+                case EdmPrimitiveTypeKind.GeometryCollection:
+                case EdmPrimitiveTypeKind.GeometryMultiPolygon:
+                case EdmPrimitiveTypeKind.GeometryMultiLineString:
+                case EdmPrimitiveTypeKind.GeometryMultiPoint:
+                    return this.ProcessSpatialTypeReferenceAsync(reference.AsSpatial());
+                case EdmPrimitiveTypeKind.Boolean:
+                case EdmPrimitiveTypeKind.Byte:
+                case EdmPrimitiveTypeKind.Double:
+                case EdmPrimitiveTypeKind.Guid:
+                case EdmPrimitiveTypeKind.Int16:
+                case EdmPrimitiveTypeKind.Int32:
+                case EdmPrimitiveTypeKind.Int64:
+                case EdmPrimitiveTypeKind.SByte:
+                case EdmPrimitiveTypeKind.Single:
+                case EdmPrimitiveTypeKind.Stream:
+                case EdmPrimitiveTypeKind.Date:
+                case EdmPrimitiveTypeKind.PrimitiveType:
+                case EdmPrimitiveTypeKind.None:
+                    return this.ProcessPrimitiveTypeReferenceAsync(reference);
+                default:
+                    throw new InvalidOperationException(Edm.Strings.UnknownEnumVal_PrimitiveKind(reference.PrimitiveKind().ToString()));

```

Comment:
See my reasoning elsewhere https://github.com/OData/odata.net/pull/3081/files#r1796808438

Reply:
Resolved by using Contract.Assert and return faulted task:
```cs
...
 Contract.Assert(false, Edm.Strings.UnknownEnumVal_ExpressionKind(expression.ExpressionKind));
 return Task.FromException<InvalidOperationException>(new InvalidOperationException(Edm.Strings.UnknownEnumVal_ExpressionKind(expression.ExpressionKind)));
...
```